velero-csi-plugin/volumesnapshotcontents_restorer.go

```diff
@@ -39,35 +40,45 @@ func (p *VSCRestorer) AppliesTo() (velero.ResourceSelector, error) {
 }
 
 func (p *VSCRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
-	p.log.Info("Starting VSCRestorer")
+	p.log.Info("Executing VSCRestorerAction")
 	var vsc snapshotv1beta1api.VolumeSnapshotContent
 
 	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(input.Item.UnstructuredContent(), &vsc); err != nil {
-		return &velero.RestoreItemActionExecuteOutput{}, err
+		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.Item from unstructured")
 	}
 
-	_, snapshotClient, err := getClients()
-	if err != nil {
-		return nil, errors.WithStack(err)
+	var vscFromBackup snapshotv1beta1api.VolumeSnapshotContent
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured((input.ItemFromBackup.UnstructuredContent()), &vscFromBackup); err != nil {
+		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.ItemFromBackup from unstructured")
 	}
 
-	snapRef := vsc.Spec.VolumeSnapshotRef
-	volumeSnapshot, err := snapshotClient.SnapshotV1beta1().VolumeSnapshots(snapRef.Namespace).Get(snapRef.Name, metav1.GetOptions{})
-	//TODO: better error handling, account for a 404
-	if err != nil {
-		return nil, errors.WithStack(err)
+	if vscFromBackup.Status == nil || vscFromBackup.Status.SnapshotHandle == nil {
+		return &velero.RestoreItemActionExecuteOutput{}, errors.Errorf("unable to lookup snapshotHandle from status")
 	}
 
-	// Make sure we're referencing the new UID for the restored VolumeSnapshot
-	// This is likely unnecessary, but what _is_ necessary is that the old UID be cleared.
-	vsc.Spec.VolumeSnapshotRef.UID = volumeSnapshot.UID
+	vscSnapshotHandle := *vscFromBackup.Status.SnapshotHandle
+	vscSnapshotClass := *vsc.Spec.VolumeSnapshotClassName
+	vsc.Spec = snapshotv1beta1api.VolumeSnapshotContentSpec{
```

Comment:
not clear on why we're needing to re-instantiate the entire spec - can you add some comments to clarify what's going on here?

Reply:
As far as I can tell, the specific things we're doing here are:
- set the `Source` field to contain the `SnapshotHandle`
- remove the old VolumeSnapshot UID from the `VolumeSnapshotRef` field

If that's the case, I think this is clearer and less error-prone if we just make those modifications, rather than re-instantiating the whole spec.

Will see what @nrb thinks here as well.

Reply:
I agree that by re-instantiating the spec, we might drop fields that may be necessary. I've fixed this to avoid re-instantiating the entire spec. 
Going to mark this as resolved. feel free to unresolve it if needed.